packages/network/src/v2/mode/syncTablesFromMode.ts

```diff
@@ -0,0 +1,82 @@
+import { ComponentValue } from "@latticexyz/recs";
+import { AbiTypeToSchemaType, encodeSchema } from "@latticexyz/schema-type";
+import { QueryLayerClient } from "@latticexyz/services/protobuf/ts/mode/mode";
+import { arrayToHex, TableId } from "@latticexyz/utils";
+import { Contract } from "ethers";
+import { NetworkEvents } from "../../types";
+
+import { CacheStore, createCacheStore, storeEvent } from "../../workers";
+import { keyTupleToEntityID } from "../keyTupleToEntityID";
+import { decodeValue } from "../schemas/decodeValue";
+import { registerMetadata } from "../schemas/tableMetadata";
+import { registerSchema } from "../schemas/tableSchemas";
+import { getBlockNumberFromModeTable } from "./getBlockNumberFromModeTable";
+
+export async function syncTablesFromMode(
+  client: QueryLayerClient,
+  chainId: number,
+  world: Contract,
+  setPercentage?: (progress: number) => void
+): Promise<CacheStore> {
+  const cacheStore = createCacheStore();
+
+  const response = await client.getState({
+    chainTables: [],
+    worldTables: [],
+    namespace: {
+      chainId: chainId.toString(),
+      worldAddress: world.address,
+    },
+  });
+  console.log("syncTablesFromMode", response);
+
+  const numRowsTotal = Object.values(response.worldTables).reduce((sum, table) => sum + table.rows.length, 0);
+  let numRowsProcessed = 0;
+
+  const blockNumber = getBlockNumberFromModeTable(response.chainTables["block_number"]);
+  const registrationPromises: Promise<unknown>[] = [];
+
+  for (const [fullTableName, { rows, cols, types }] of Object.entries(response.worldTables)) {
+    const [tableNamespace, tableName] = fullTableName.split("__");
+    const tableId = new TableId(tableNamespace, tableName);
+
+    const component = tableId.toString();
+
+    // TODO: separate keys and values/fields in MODE, but we'll infer for now
+    const keyLength = cols.findIndex((col) => !col.startsWith("key_"));
+    const keyNames = cols.slice(0, keyLength); // TODO: key names are not currently included in the MetadataTable, so these are unused
+    const keyTypes = types.slice(0, keyLength).map((abiType) => AbiTypeToSchemaType[abiType]);
+    const fieldNames = cols.slice(keyLength);
+    const fieldTypes = types
+      .slice(keyLength)
+      .map((modeType) => modeType.match(/tuple\((.*)\[]\)/)?.[1] ?? modeType) // TODO: remove this hack once MODE is fixed
```

Comment:
Context: MODE currently returns an invalid `fieldType` value for dynamic types. We manually extract the valid type here until the issue is fixed in MODE.

![image](https://user-images.githubusercontent.com/89248902/229104639-2e3141af-38a5-4005-a9c9-fe0fc19997b9.png)